### PR TITLE
 mintro: Allow introspect --projectinfo without build directory

### PR DESCRIPTION
**Pull request by trilader**
This variant was added to allow introspection before configuring a build
directory. This is useful for IDE integration to allow displaying and/or
setting options for the initial configuration of the build directory.

It also allows showing basic information about the project even if it's
not yet configured or configuring failed.

The project 'name' field in --projectinfo is used inconsistently:
For the top level project it always shows the name configured in
the top level meson.build file. For subprojects it's referring to the
name of the directory the subproject's meson.build is contained in.

To have a consistent output and preserve the existing behavior this adds
the 'descriptive_name' field which always shows the name set in the
project.

To be consistent the 'descriptive_name' field was also added to the
--projectfiles variant that uses an already configured build.

It also extends the information shown with the list of buildsystem-files.
This is currently only implemented in the variant for unconfigured
projects.

### CODE REVIEW

**Comment by jpakkane** [comment]
We are currently in dev freeze for 0.48.0 during which we won't merge new features. This MR will be processed once the release is out. Thanks.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

This also needs a test. It should run the introspection command on one of the test case dirs that has a subdirectory with options and check:

 - that global options are shown
 - that top level project options are shown
 - that subproject project options are shown

**Comment by jpakkane** [comment]
And since this is new functionality, there should be a release note snippet.

File: run_unittests.py [hunk 1]
```diff
@@ -2746,6 +2753,39 @@ def test_buildtype_setting(self):
...
+                              })
+
+        testfile = os.path.join(self.common_test_dir, '44 options', 'meson.build')
+        res = self.introspect_directory(testfile, '--projectinfo')
+        self.assertEqual(res, {'buildsystem_files': ['meson_options.txt', 'meson.build'],
+                               'name': 'options',
+                               'version': None,
+                               'descriptive_name': 'options',
+                               'subprojects': []
+                              })
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

File: run_unittests.py [hunk 2]
```diff
@@ -2746,6 +2753,39 @@ def test_buildtype_setting(self):
...
+        self.assertEqual(res, {'buildsystem_files': ['meson_options.txt', 'meson.build'],
+                               'name': 'suboptions',
+                               'version': None,
+                               'descriptive_name': 'suboptions',
+                               'subprojects': [
+                                   {'buildsystem_files': ['subprojects/subproject/meson_options.txt', 'subprojects/subproject/meson.build'],
+                                    'name': 'subproject',
+                                    'version': 'undefined',
+                                    'descriptive_name': 'subproject'}
+                                   ]
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[E123] closing bracket does not match indentation of opening bracket's line

File: run_unittests.py [hunk 3]
```diff
@@ -2746,6 +2753,39 @@ def test_buildtype_setting(self):
...
 
+    def test_introspect_projectinfo_without_configured_build(self):
+        testfile = os.path.join(self.common_test_dir, '36 run program', 'meson.build')
+        res = self.introspect_directory(testfile, '--projectinfo')
+        self.assertEqual(res, {'buildsystem_files': ['meson.build'],
+                               'name': 'run command',
+                               'version': None,
+                               'descriptive_name': 'run command',
+                               'subprojects': []
+                              })
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

File: run_unittests.py [hunk 4]
```diff
@@ -2746,6 +2753,39 @@ def test_buildtype_setting(self):
...
+                               'name': 'suboptions',
+                               'version': None,
+                               'descriptive_name': 'suboptions',
+                               'subprojects': [
+                                   {'buildsystem_files': ['subprojects/subproject/meson_options.txt', 'subprojects/subproject/meson.build'],
+                                    'name': 'subproject',
+                                    'version': 'undefined',
+                                    'descriptive_name': 'subproject'}
+                                   ]
+                              })
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

**Comment by trilader** [comment]
> This also needs a test. It should run the introspection command on one of the test case dirs that has a subdirectory with options and check:
> 
> * that global options are shown
> * that top level project options are shown
> * that subproject project options are shown

I've added tests that also include subprojects with options. 

Currenly option files are included in `buildsystem_files` but this PR does not replace `introspect --buildoptions`. Figuring out how to get the build options (in an IDE friendly format) without a configured build directory is a topic for future work.

**Comment by jpakkane** [comment]
Test failure seems relevant:

```
- {'buildsystem_files': ['meson.build', 'meson_options.txt'],
+ {'buildsystem_files': ['meson_options.txt', 'meson.build'],
```

**Comment by trilader** [comment]
I've switched the comparison of the buildsystem_files list to a set comparison so it now works regardless of the order the files are found on the disk.

**Comment by trilader** [comment]
@jpakkane Can you have a look at this now that the issues you mentioned previously are fixed?

**Comment by nirbheek** [comment]
Am I reading the examples correctly, that `meson introspect --projectinfo` won't implicitly read the top-level meson build file in the current directory?

**Comment by trilader** [comment]
That is correct.

**Comment by nirbheek** [comment]
Maybe it should? All the other meson commands that operate on the top-level build file can be run in the top of the source dir without having to explicitly give the path to the build file. If you run from a different directory, you can give the source directory as an option, too. So, f.ex.,

`meson introspect --projectinfo <srcdir>` should be the same as `meson introspect --projectinfo <srcdir>/meson.build`, and

`meson introspect --projectinfo ./meson.build` should be the same as `meson introspect --projectinfo`.

**Comment by trilader** [comment]
`meson introspect --projectinfo` can now be used in "unconfigured build" mode without explicitly specifiying `meson.build` as part of the parameter for both the implicit `.` directory as well as explicitly specified directories.

**Comment by nirbheek** [comment]
> `meson introspect --projectinfo` can now be used in "unconfigured build" mode without explicitly specifiying `meson.build` as part of the parameter for both the implicit `.` directory as well as explicitly specified directories.

Looks like the unit test is missing these cases? Could you please add that? It'll almost certainly break in the future otherwise :)

**Comment by trilader** [comment]
I've updated the test cases with a call to an unconfigured project with implicit meson.build argument.

**Comment by jpakkane** [comment]
The code says that name and descriptive name are different for subprojects, however in all examples and tests they are the same string. There should probably be a test or doc update where these two are different testing that the naming is done correctly.

Alternatively we may even consider changing the json format to avoid this issue. It is only used by Gnome Builder currently (AFAIK) and I'm not sure they use this piece of data for anything, but we would obviously need to check if anyone else is using it.

**Comment by trilader** [comment]
I've adapted the unit test in the last commit to address the different name in the project() function of a subproject.

We could just remove the name field in the top level project output. Can you CC someone from gnome-builder who can give some input on that proposal? But I don't think it's necessary to remove the field.

**Comment by jpakkane** [comment]
Just got confirmation that GB is not using that. Which means that probably nobody else is using it either and we can change it.

**Comment by trilader** [comment]
Done

**Comment by mensinda** [comment]
Azure will fail because of #4549. (Sorry my bad)

**Pull request merged**